### PR TITLE
589 column encryption

### LINKED ISSUE #589: column encryption
**Is your feature request related to a problem? Please describe.**
As a user I want to encrypt data in specific columns.

**Describe the solution you'd like**
Using a basic Java AES encryption, applied on a list of specified columns. The key should be handled via the typical credential mechanism (CLEAR, ENV, ...)

### PR DESCRIPTION
**Pull request by mand35**
### What changes are included in the pull request?
Add a new transformer, which encrypts a list of columns in the dataFrame. 
Furthermore, the counterpart, the decryptor is implemented. Both use java cipher with. 

### Why are the changes needed?
Data security becomes more and more important. Often very sensitive data is mixed with data necessary to be analysed. Therefore, we need to hide the sensitive data by encrypting it

### Aspects to consider and to be discussed
We use a GCM method with initial vector (IV), which is stored at the beginning of the encrypted message. Therewith cells with the same value result in different ciphertexts.

### CODE REVIEW

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import org.apache.spark.sql.expressions.UserDefinedFunction
+
+import java.security.SecureRandom
+import java.util.Base64
+import javax.crypto.spec.{GCMParameterSpec, SecretKeySpec}
+import javax.crypto.{Cipher, SecretKey}
+import org.apache.spark.sql.functions.{col, udf}
+
+trait EncryptDecrypt {
+  def key: Array[Byte] //= "test234".getBytes
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala]: please remove the confusing comment

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  private val secureRandom = new SecureRandom()
+
+
+  def process(df: DataFrame, encryptColumns: Seq[String]): DataFrame = {
+    var dfEnc = df
+    for (colName <- encryptColumns) {
+      dfEnc = dfEnc.withColumn(colName, cryptUDF(col(colName)))
+    }
+    dfEnc
+  }
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala]: ``
```suggestion
  def process(df: DataFrame, encryptColumns: Seq[String]): DataFrame = {
   
   encryptColumns.foldLeft(df){
   case (dfTemp, colName) => dfTemp.withColumn(colName, cryptUDF(col(colName)))
   }
  }
```
Try this! FoldLeft to the rescue. You can replace every loop with foldLefts. Complex and hierarchical loops get easier to read and you don t need evil variables

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    // Run SDLB
+    implicit val hadoopConf: Configuration = session.sparkContext.hadoopConfiguration
+    val initialSubFeeds: Seq[SparkSubFeed] = Seq(SparkSubFeed(None, srcDO.id, Seq()))
+    val (subFeeds, stats) = sdlb.exec(sdlConfig, SDLExecutionId.executionId1, runStartTime = LocalDateTime.now, attemptStartTime = LocalDateTime.now, actionsToSkip = Map(), initialSubFeeds = initialSubFeeds, dataObjectsState = Seq(), stateStore = None, stateListeners = Seq(), simulation = false, globalConfig = globalConfig)
+
+    // check result
+    val tgt = instanceRegistry.get[ParquetFileDataObject]("tgt")
+    val dfTgt = tgt.getSparkDataFrame()
+    val colName = dfTgt.columns
+    //assert(colName.toSeq == Seq("testcolumn", "secondcolumn", "thirdcolumn"))
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: hmm is this test not working or why is the assert commented out?

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  test("test column decryption") {
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: looks like the second test relies on the output of the first. Why not put them into the same test, with 2 asserts? I think it's readable enough

**Review by Geheiner — CHANGES_REQUESTED** [review]
Nice feature, nice tests! I do have some comments. Especially, Try out foldLeft ;)

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    val srcDO = instanceRegistry.get[CsvFileDataObject]("src")
+    assert(srcDO != None)
+    val dfSrc = Seq(("testData", "Foo", "ice"), ("bar", "Space", "water"), ("gogo", "Space", "water")).toDF("c1", "c2", "c3")
+    srcDO.writeDataFrame(SparkDataFrame(dfSrc), Seq())(TestUtil.getDefaultActionPipelineContext(sdlb.instanceRegistry))
+
+    // Run SDLB
+    implicit val hadoopConf: Configuration = session.sparkContext.hadoopConfiguration
+    val initialSubFeeds: Seq[SparkSubFeed] = Seq(SparkSubFeed(None, srcDO.id, Seq()))
+    val (subFeeds, stats) = sdlb.exec(sdlConfig, SDLExecutionId.executionId1, runStartTime = LocalDateTime.now, attemptStartTime = LocalDateTime.now, actionsToSkip = Map(), initialSubFeeds = initialSubFeeds, dataObjectsState = Seq(), stateStore = None, stateListeners = Seq(), simulation = false, globalConfig = globalConfig)
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: you could also use a method called startSimulationFromConfigFile where you can directly pass the dataframe so you don t need to write to a file

**Comment by Geheiner** [comment]
> We use a GCM method with initial vector (IV), which is stored at the beginning of the encrypted message. Therewith cells with the same value result in different ciphertexts.

For me this sounds good as it makes reverse engineering the encrypted data harder

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala (hunk 1 above)
Comment (mand35) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala]: removed, thanks

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala (hunk 3 above)
Comment (mand35) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: corrected the assert, thanks.

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala (hunk 4 above)
Comment (mand35) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: Indeed, better to merge to also remove the dependencies between test cases. Thanks.

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala (hunk 5 above)
Comment (mand35) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: would be maybe an option.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  private val IV_SIZE = 128
+  private val TAG_BIT_LENGTH = 128
+
+  private val secureRandom = new SecureRandom()
+
+
+  def process(df: DataFrame, encryptColumns: Seq[String]): DataFrame = {
+   
+   encryptColumns.foldLeft(df){
+   case (dfTemp, colName) => dfTemp.withColumn(colName, cryptUDF(col(colName)))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformer.scala]: can you run intelliJ code formatter?

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        |     inputId = enc
+        |     outputId = dec
+        |     metadata {
+        |       feed = test_run
+        |     }
+        |     transformers = [{
+        |       type = DecryptColumnsTransformer
+        |       decryptColumns = ["c2","c3"]
+        |       keyVariable = "CLEAR#A%D*G-KaPdSgVkYp"
+        |       #keyVariable = "CLEAR#keyblabla"
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/generic/transformer/EncryptColumnsTransformerTest.scala]: remove unnecessary line

**Review by Geheiner — APPROVED** [review]
LGTM! I have 2 minor comments

**Review by zzeekk — APPROVED** [review]

**Pull request merged**